File: MPP-firma-transport-java/FirmaNetworking/src/main/java/org/example/FirmaServicesRPCProxy.java
package org.example;

import org.example.DTO.FilterDTO;

import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class FirmaServicesRPCProxy implements IFirmaServices{
    private String host;
    private int port;

    private IFirmaObserver client;
    private ObjectInputStream input;
    private ObjectOutputStream output;
    private Socket connection;
    private volatile boolean finished;
    private BlockingQueue<Response> queueResponses;

    public FirmaServicesRPCProxy(String host, int port) {
        this.host = host;
        this.port = port;
        queueResponses = new LinkedBlockingQueue<>();
    }

    @Override
    public Optional<Oficiu> login(String username, String password, IFirmaObserver client) throws AppException {
        initializeConnection();
        Oficiu utilizator = new Oficiu(1L,"",username, password);
        Request request = new Request.Builder().type(RequestType.LOGIN).data(utilizator).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.OK) {
            this.client = client;
            return Optional.of((Oficiu) response.getData());
        }
        if (response.getType() == ResponseType.ERROR) {
            String error = response.getData().toString();
            closeConnection();
            throw new AppException(error);
        }
        return Optional.empty();
    }

    @Override
    public void logout(String username) throws AppException {
        Request request = new Request.Builder().type(RequestType.LOGOUT).data(username).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR) {
            String error = response.getData().toString();
            throw new AppException(error);
        }
    }

    private void closeConnection() {
        finished = true;
        try {
            input.close();
            output.close();
            connection.close();
            client = null;
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void sendRequest(Request request) throws AppException {
        try {
            output.writeObject(request);
            output.flush();
        } catch (Exception e) {
            throw new AppException("Error sending object " + e);
        }
    }

    private Response readResponse() {
        Response response = null;
        try {
            response = queueResponses.take();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return response;
    }

    private void initializeConnection() throws AppException {
        try {
            connection = new Socket(host, port);
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            finished = false;
            startReader();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void startReader() {
        Thread tw = new Thread(new ReaderThread());
        tw.start();
    }
    @Override
    public List<CursaDTO> getAllCurse() throws AppException {
        Request request = new Request.Builder().type(RequestType.GET_Curse).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR) {
            String error = response.getData().toString();
            throw new AppException(error);
        }
        return (List<CursaDTO>) response.getData();
    }

    @Override
    public List<CursaDTO> getCursaByDestinatieData(String destinatie, String data) throws AppException {
        Request request = new Request.Builder().type(RequestType.GET_Curse_FILTERED).data(new FilterDTO(destinatie,data)).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR) {
            String error = response.getData().toString();
            throw new AppException(error);
        }
        return (List<CursaDTO>) response.getData();
    }

    @Override
    public void makeReservation(Long cursaID, String numeClient, Integer nrLocuri) throws AppException {
        var rezervare = new RezervareDTO(1L,numeClient,nrLocuri,cursaID);
        Request request = new Request.Builder().type(RequestType.RESERVATION).data(rezervare).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR) {
            String error = response.getData().toString();
            throw new AppException(error);
        }
    }

    @Override
    public List<LocDTO> getLocuriDisponibile(Long idCursa) throws AppException {
        Request request = new Request.Builder().type(RequestType.GET_LOCURI).data(idCursa).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR) {
            String error = response.getData().toString();
            throw new AppException(error);
        }
        return (List<LocDTO>) response.getData();
    }

    @Override
    public RezervareDTO getRezervareById(Long rezervare) throws AppException{
        Request request = new Request.Builder().type(RequestType.GET_REZERVARE).data(rezervare).build();
        sendRequest(request);
        Response response = readResponse();
        if (response.getType() == ResponseType.ERROR) {
            return null;
        }
        return (RezervareDTO) response.getData();



    }

    private void handleUpdate(Response response) {
        if (response.getType() == ResponseType.UPDATE) {
            client.reservationUpdate();
        }
    }
    private boolean isUpdateResponse(Response response) {
        return response.getType() == ResponseType.UPDATE;
    }

    private class ReaderThread implements Runnable {
        public void run() {
            while (!finished) {
                try {
                    Object response = input.readObject();
                    System.out.println("Response received " + response);
                    if (response instanceof Response) {
                        Response response1 = (Response) response;
                        if (isUpdateResponse(response1)) {
                            handleUpdate(response1);
                        } else {
                            try {
                                queueResponses.put((Response) response);
                            } catch (InterruptedException e) {
                                e.printStackTrace();
                            }
                        }
                    }
                } catch (Exception e) {
                    System.out.println("Reading error " + e);
                }
            }
        }
    }


}

File: mpp-firma-transport-c#/FirmaNetworking/FirmaServicesRPCProxy.cs
﻿using FirmaNetworking.DTO;
using FirmaNetworking;
using FirmaService;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using FirmaModel1;
using FirmaNetworking.DTO;
using FirmaService;
using Utils;

namespace FirmaNetworking
{
    public class FirmaServicesRPCProxy : IFirmaServices
    {
        private string host;
        private int port;
        private IFirmaObserver client;
        private TcpClient connection;
        private NetworkStream stream;
        private IFormatter formatter;
        private volatile bool finished;
        private BlockingCollection<Response> queueResponses;

        public FirmaServicesRPCProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            // formatter = new BinaryFormatter();
            queueResponses = new BlockingCollection<Response>();
        }

        private void InitializeConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                finished = false;
                StartReader();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private void StartReader()
        {
            Thread tw = new Thread(Run);
            tw.Start();
        }

        public virtual void Run()
        {
            while (!finished)
            {
                try
                {
                    if (stream.CanRead && stream.DataAvailable)
                    {
                        // Console.WriteLine(stream);
                        // object raspuns = formatter.Deserialize(stream);
                        // stream.Flush();
                        // stream.Position = 0L;
                        Response response = (Response)formatter.Deserialize(stream);
                        Console.WriteLine("Response received " + response);
                        if (IsUpdate(response))
                        {
                            HandleUpdate(response);
                        }
                        else
                        {
                            queueResponses.Add(response);
                        }
                    }
                    else
                    {
                        Thread.Sleep(1);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                    CloseConnection();
                    finished = true;
                }
            }
        }

        private void CloseConnection()
        {
            finished = true;
            try
            {
                stream.Close();
                connection.Close();
                client = null;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }


        public Optional<Oficiu> login(string username, string password, IFirmaObserver client)
        {
            InitializeConnection();
            Oficiu utilizator = new Oficiu("", username, password);
            Request request = new Request.Builder().SetType(RequestType.LOGIN).SetData(utilizator).Build();
            SendRequest(request);
            Response response = ReadResponse();
            if (response.Type == ResponseType.OK)
            {
                this.client = client;
                return new Optional<Oficiu>((Oficiu)response.Data);
            }
            if (response.Type == ResponseType.ERROR)
            {
                string message = (string)response.Data;
                CloseConnection();
                throw new AppException(message);
            }
            return Optional<Oficiu>.Empty();
        }

        private Response ReadResponse()
        {
            Response response = null;
            // int i = 10;
            try
            {
                response = queueResponses.Take();
                // incearca:
                // if (queueResponses.TryDequeue(out Response topElement))
                // {
                //     response = topElement;
                //     Console.WriteLine("Element removed: {0}", topElement);
                //     return response;
                // }
                //
                // if (i >= 0)
                // {
                //
                //     i--;
                //     Thread.Sleep(100);
                //     goto incearca;
                // }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }

            return response;
        }

        private void SendRequest(Request request)
        {
            lock (stream)
            {
                try
                {
                    formatter.Serialize(stream, request);
                    stream.Flush();
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
        }

        public void logout(string username)
        {
            Request request = new Request.Builder().SetType(RequestType.LOGOUT).SetData(username).Build();
            SendRequest(request);
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                string message = (string)response.Data;
                throw new AppException(message);
            }
            // this.CloseConnection();
        }

        public List<CursaDTO> getAllCurse()
        {
            Request request = new Request.Builder().SetType(RequestType.GET_Curse).Build();
            SendRequest(request);
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                string message = (string)response.Data;
                throw new AppException(message);
            }
            return (List<CursaDTO>)response.Data;
        }

        public List<CursaDTO> getCursaByDestinatieData(string destinatie, string data)
        {
            Request request = new Request.Builder().SetType(RequestType.GET_Curse_FILTERED).SetData(new FilterDTO(destinatie, data)).Build();
            SendRequest(request);
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                string message = (string)response.Data;
                throw new AppException(message);
            }
            return (List<CursaDTO>)response.Data;
        }

        public void makeReservation(long idCursa, String numeClient, int nrLocuri)
        {
            RezervareDTO rezervare = new RezervareDTO(1L, numeClient, nrLocuri, idCursa);
            Request request = new Request.Builder().SetType(RequestType.RESERVATION).SetData(rezervare).Build();
            SendRequest(request);
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                string message = (string)response.Data;
                throw new AppException(message);
            }
        }

        public List<LocDTO> getLocuriDisponibile(long idCursa)
        {

            Request request = new Request.Builder().SetType(RequestType.GET_Locuri).SetData(idCursa).Build();
            SendRequest(request);
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                string message = (string)response.Data;
                throw new AppException(message);
            }
            return (List<LocDTO>)response.Data;
        }

        public RezervareDTO getRezervareById(long id)
        {
            Request request = new Request.Builder().SetType(RequestType.Get_Rezervare).SetData(id).Build();
            SendRequest(request);
            Response response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                string message = (string)response.Data;
                throw new AppException(message);
            }
            return (RezervareDTO)response.Data;
        }

        private bool IsUpdate(Response response)
        {
            return response.Type == ResponseType.UPDATE;
        }

        private void HandleUpdate(Response response)
        {
            if (response.Type == ResponseType.UPDATE)
            {
                client.ReservationUpdate();
            }
        }
    }
}